connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java

```diff
@@ -151,4 +160,19 @@ public void run() {
         }
     }
 
+    public boolean isPaused() {
+        return this.targetState.get() == TargetState.PAUSED;
+    }
+
+    public void transitionTo(TargetState state) {
+        TargetState oldState = this.targetState.getAndSet(state);
+        if (state != oldState && running.get()) {
+            // FIXME: this is racy
```

Comment:
Ah, I forgot about this. The problem is that the task could fail at the same time that it is supposed to transition, which could cause the task to be seen incorrectly as paused instead of failed. So we need to protect the order of operations on the status listener.